### ISSUE TITLE
zdharma performance optimizations

### ISSUE THREAD

**Issue by danielshahaf** [zsh-users/zsh-syntax-highlighting#434]
http://zdharma.org/posts/Fast-Syntax-Highlighting-Released

- [ ] Check license / permission to use that
- [ ] Check if there's already a PR for these.
- [ ] Review and cherry pick.

**Comment by maximbaz** [comment]
The [zdharma/fast-syntax-hightlighting/LICENSE](https://github.com/zdharma/fast-syntax-highlighting/blob/de8d9558fc34e67334a73716eec87bc5afbe2e6a/LICENSE) is the same as in this project, even with exactly the same copyright notice. The license requires preserving the copyright and disclaimer, both items are already fulfilled.

**Comment by Aeon** [comment]
+1

**Comment by danielshahaf** [comment]
Looking at the copyright notices:

- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/LICENSE#L1-L2
- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/fast-syntax-highlighting.plugin.zsh#L1-L26
- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/themes/q-jmnemonic.ini#L5-L7
- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/themes/default.ini
- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/:chroma/-autoload.ch#L2
- https://github.com/zdharma/fast-syntax-highlighting/blob/303eeee81859094385605f7c978801748d71056c/:chroma/-subversion.ch#L2-L29

It's not clear to me under what license https://github.com/zdharma/fast-syntax-highlighting is under.

@psprint Would you be so kind as to clarify whether or not the z-sy-h project has permission to cherry-pick code from https://github.com/zdharma/fast-syntax-highlighting?  What license is that repository under?

Cheers!

**Comment by danielshahaf** [comment]
Pinged again by email:

```
From: Daniel Shahaf
To: Sebastian Gniazdowski
Cc: Matthew Martin
Subject: Fw: [zsh-users/zsh-syntax-highlighting] zdharma performance  optimizations (#434)
Date: Sun, 29 Mar 2020 22:31:44 +0000
Message-ID: <20200329223144.70d00614@tarpaulin.shahaf.local2>
```

**Comment by danielshahaf** [comment]
For clarity: under copyright law, code may not be reused unless its author (as copyright holder) specifies otherwise.  (It makes no difference that the code in question is public and derived from BSD-licensed code.)  In our case, that means z-sy-h may not use any piece of code that's not explicitly stated to be BSD-licensed, barring specific permission from the author of the piece of code in question.  Thus, the query to Sebastian regarding the license of the code he's distributing, and whether z-sy-h may use that code under a BSD license.

**Comment by psprint** [comment]
Yes, you are allowed to reuse the code, the license is the same.

**Comment by danielshahaf** [comment]
Thanks, Sebastian!

**Comment by danielshahaf** [comment]
Folks — if there are particular features you'd like to propose to be cherry-picked, please file a new bug report for each such feature.  (In order to avoid duplicates, please reference this issue #434 in the new bug reports.)  Cheers!

**Comment by danielshahaf** [comment]
The "HSMW" prefix in the names of some variables, early in the history of `fast-highlight` in https://github.com/zdharma/fast-syntax-highlighting, refers to https://github.com/zdharma/history-search-multi-word.

**Comment by walles** [comment]
zdharma has moved here, the old project seems to have been taken down: https://github.com/zdharma-continuum/fast-syntax-highlighting
